File: benchmarks/python/run_latency.py
"""
Python Latency Benchmark — prompt-sanitizer FAST / SMART / FULL
================================================================

Usage:
    pip install -r requirements.txt
    python run_latency.py

Measures:
  - Median, p95, p99 latency in ms for 1000 iterations per mode
  - Throughput in calls/sec
  - Optionally compares against Presidio

Benchmark text: a 3-sentence paragraph containing multiple PII types.
"""
from __future__ import annotations

import asyncio
import statistics
import sys
import time
from pathlib import Path

# ---------------------------------------------------------------------------
# Benchmark text
# ---------------------------------------------------------------------------

SAMPLE_SHORT = "My email is alice@example.com and my SSN is 078-05-1120."

SAMPLE_MEDIUM = (
    "Hi, my name is Alice Walker. "
    "You can reach me at alice@example.com or call (415) 867-5309. "
    "My SSN is 078-05-1120 and I last used card 4111 1111 1111 1111 "
    "for a payment to account GB29 NWBK 6016 1331 9268 19. "
    "Our server runs at 192.168.1.105."
)

SAMPLE_LONG = SAMPLE_MEDIUM * 5  # ~5× medium

WARMUP_ITERS = 10
BENCH_ITERS  = 500


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def stats(times_ms: list[float]) -> dict:
    times_ms = sorted(times_ms)
    n = len(times_ms)
    return {
        "min":    min(times_ms),
        "median": statistics.median(times_ms),
        "p95":    times_ms[int(0.95 * n)],
        "p99":    times_ms[int(0.99 * n)],
        "max":    max(times_ms),
        "mean":   statistics.mean(times_ms),
        "rps":    1000 / statistics.median(times_ms),
    }


def bench_sync(fn, n: int) -> list[float]:
    times = []
    for _ in range(n):
        t0 = time.perf_counter()
        fn()
        times.append((time.perf_counter() - t0) * 1000)
    return times


async def bench_async(fn, n: int) -> list[float]:
    times = []
    for _ in range(n):
        t0 = time.perf_counter()
        await fn()
        times.append((time.perf_counter() - t0) * 1000)
    return times


def print_row(label: str, s: dict, text_label: str) -> None:
    print(
        f"  {label:<38} {text_label:<8} "
        f"{s['median']:>7.2f}  {s['p95']:>7.2f}  {s['p99']:>7.2f}"
        f"  {s['rps']:>8.0f}"
    )


# ---------------------------------------------------------------------------
# Benchmark runners
# ---------------------------------------------------------------------------

sys.path.insert(0, str(Path(__file__).parent.parent.parent / "packages" / "python"))


async def bench_prompt_sanitizer() -> None:
    from prompt_sanitizer import Sanitizer, Mode  # type: ignore

    print("\n── prompt-sanitizer ─────────────────────────────────────────────────")
    print(f"  {'Tool':<38} {'Text':<8} {'median':>7}  {'p95':>7}  {'p99':>7}  {'rps':>8}")
    print(f"  {'-'*38} {'-'*8} {'-'*7}  {'-'*7}  {'-'*7}  {'-'*8}")

    for mode_name, mode in [("FAST", Mode.FAST), ("SMART", Mode.SMART), ("FULL", Mode.FULL)]:
        s = Sanitizer(mode=mode, on_detect="warn")

        for text_label, text in [("short", SAMPLE_SHORT), ("medium", SAMPLE_MEDIUM), ("long", SAMPLE_LONG)]:
            # Warmup
            for _ in range(WARMUP_ITERS):
                await s.sanitize(text)

            # Bench
            times = await bench_async(lambda t=text: s.sanitize(t), BENCH_ITERS)
            s_stats = stats(times)
            print_row(f"prompt-sanitizer  Mode.{mode_name:<5}", s_stats, text_label)


def bench_presidio() -> None:
    try:
        from presidio_analyzer import AnalyzerEngine  # type: ignore
    except ImportError:
        print("\n── Presidio ─────────────────────────────────────────────────────────")
        print("  [SKIP] presidio-analyzer not installed (pip install presidio-analyzer)")
        return

    print("\n── Presidio ─────────────────────────────────────────────────────────")
    print(f"  {'Tool':<38} {'Text':<8} {'median':>7}  {'p95':>7}  {'p99':>7}  {'rps':>8}")
    print(f"  {'-'*38} {'-'*8} {'-'*7}  {'-'*7}  {'-'*7}  {'-'*8}")

    print("  Loading AnalyzerEngine (spaCy)...", end=" ", flush=True)
    analyzer = AnalyzerEngine()
    print("ready.")

    for text_label, text in [("short", SAMPLE_SHORT), ("medium", SAMPLE_MEDIUM), ("long", SAMPLE_LONG)]:
        # Warmup
        for _ in range(WARMUP_ITERS):
            analyzer.analyze(text=text, language="en")

        # Bench
        times = bench_sync(lambda t=text: analyzer.analyze(text=t, language="en"), BENCH_ITERS)
        s_stats = stats(times)
        print_row("Presidio (spaCy)", s_stats, text_label)


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main() -> None:
    print(f"\nPython Latency Benchmark — {BENCH_ITERS} iterations per scenario")
    print(f"  All times in milliseconds. rps = calls/second at median latency.\n")

    print(f"  {'Tool':<38} {'Text':<8} {'median':>7}  {'p95':>7}  {'p99':>7}  {'rps':>8}")
    print(f"  {'='*38} {'='*8} {'='*7}  {'='*7}  {'='*7}  {'='*8}")

    asyncio.run(bench_prompt_sanitizer())
    bench_presidio()

    print("\n  Legend:")
    print("    short  = 1 sentence,  ~58 chars,  2 PII entities")
    print("    medium = 5 sentences, ~280 chars,  6 PII entities")
    print("    long   = 25 sentences,~1400 chars, 30 PII entities")
    print()


if __name__ == "__main__":
    main()

File: benchmarks/javascript/run_latency.mjs
/**
 * JS Latency Benchmark — prompt-sanitizer FAST / SMART vs OpenRedaction
 * ======================================================================
 *
 * Usage:
 *   npm install
 *   node run_latency.mjs
 *
 * Measures: min / median / p95 / p99 latency in ms + RPS at median.
 * Input sizes: short (~58 chars), medium (~280 chars), long (~1400 chars).
 */

import { performance } from "perf_hooks";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));

const WARMUP_ITERS = 10;
const BENCH_ITERS  = 300;

// ---------------------------------------------------------------------------
// Benchmark texts
// ---------------------------------------------------------------------------

const SAMPLE_SHORT = "My email is alice@example.com and my SSN is 078-05-1120.";

const SAMPLE_MEDIUM = [
  "Hi, my name is Alice Walker.",
  "You can reach me at alice@example.com or call (415) 867-5309.",
  "My SSN is 078-05-1120 and I last used card 4111 1111 1111 1111",
  "for a payment to account GB29 NWBK 6016 1331 9268 19.",
  "Our server runs at 192.168.1.105.",
].join(" ");

const SAMPLE_LONG = SAMPLE_MEDIUM.repeat(5);

// ---------------------------------------------------------------------------
// Stats helpers
// ---------------------------------------------------------------------------

function computeStats(timesMs) {
  const sorted = [...timesMs].sort((a, b) => a - b);
  const n = sorted.length;
  const sum = sorted.reduce((a, b) => a + b, 0);
  return {
    min:    sorted[0],
    median: sorted[Math.floor(n / 2)],
    p95:    sorted[Math.floor(0.95 * n)],
    p99:    sorted[Math.floor(0.99 * n)],
    max:    sorted[n - 1],
    mean:   sum / n,
    rps:    1000 / sorted[Math.floor(n / 2)],
  };
}

async function benchAsync(fn, n) {
  const times = [];
  for (let i = 0; i < n; i++) {
    const t0 = performance.now();
    await fn();
    times.push(performance.now() - t0);
  }
  return times;
}

function printHeader() {
  const label = "  " + "Tool".padEnd(36) + " " + "Text".padEnd(8);
  const cols  = " median    p95    p99    rps";
  console.log(label + cols);
  console.log("  " + "-".repeat(36) + " " + "-".repeat(8) + " " + "-".repeat(28));
}

function printRow(label, textLabel, s) {
  const l = ("  " + label).padEnd(38);
  const t = textLabel.padEnd(8);
  const row = [s.median, s.p95, s.p99].map((v) => v.toFixed(2).padStart(7)).join(" ");
  const rps = String(Math.round(s.rps)).padStart(7);
  console.log(l + " " + t + " " + row + " " + rps);
}

// ---------------------------------------------------------------------------
// Runners
// ---------------------------------------------------------------------------

async function loadPromptSanitizer() {
  return import(
    resolve(__dirname, "../../packages/javascript/src/index.js")
  ).catch(() =>
    import(resolve(__dirname, "../../packages/javascript/dist/index.js"))
  );
}

async function benchPromptSanitizer() {
  const { Sanitizer, Mode } = await loadPromptSanitizer();

  console.log("\n── prompt-sanitizer ──────────────────────────────────────────────────");
  printHeader();

  for (const [modeName, mode] of [["FAST", Mode.FAST], ["SMART", Mode.SMART]]) {
    const s = new Sanitizer({ mode, onDetect: "warn" });

    for (const [textLabel, text] of [["short", SAMPLE_SHORT], ["medium", SAMPLE_MEDIUM], ["long", SAMPLE_LONG]]) {
      // Warmup
      for (let i = 0; i < WARMUP_ITERS; i++) await s.sanitize(text);

      const times = await benchAsync(() => s.sanitize(text), BENCH_ITERS);
      printRow(`prompt-sanitizer  ${modeName}`, textLabel, computeStats(times));
    }
  }
}

async function benchOpenRedaction() {
  let OpenRedaction;
  try {
    ({ OpenRedaction } = await import("openredaction"));
  } catch {
    console.log("\n── OpenRedaction ──────────────────────────────────────────────────────");
    console.log("  [SKIP] openredaction not installed (npm install openredaction)");
    return;
  }

  console.log("\n── OpenRedaction ──────────────────────────────────────────────────────");
  printHeader();

  const redactor = new OpenRedaction();

  for (const [textLabel, text] of [["short", SAMPLE_SHORT], ["medium", SAMPLE_MEDIUM], ["long", SAMPLE_LONG]]) {
    for (let i = 0; i < WARMUP_ITERS; i++) await redactor.detect(text);
    const times = await benchAsync(() => redactor.detect(text), BENCH_ITERS);
    printRow("OpenRedaction (regex)", textLabel, computeStats(times));
  }
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main() {
  console.log(`\nJS Latency Benchmark — ${BENCH_ITERS} iterations per scenario`);
  console.log("  All times in milliseconds. rps = calls/second at median.\n");
  console.log("  " + "=".repeat(70));
  console.log(`  ${"Tool".padEnd(36)} ${"Text".padEnd(8)} ${"median":>7}  ${"p95":>7}  ${"p99":>7}  ${"rps":>7}`);
  console.log("  " + "=".repeat(70));

  await benchPromptSanitizer();
  await benchOpenRedaction();

  console.log("\n  Legend:");
  console.log("    short  = 1 sentence,  ~58 chars,  2 PII entities");
  console.log("    medium = 5 sentences, ~280 chars,  6 PII entities");
  console.log("    long   = 25 sentences,~1400 chars, 30 PII entities\n");
}

main().catch(console.error);
